verification_service/main.py

```diff
@@ -0,0 +1,309 @@
+
+        processed_events_total.inc()
+
+        user_email_address = None
+        user_verified = False
+
+        # Determine how to get the user email address based on event type
+        if (
+            event_type
+            in project_events + user_events + issue_events + issue_note_events
+        ):
+            user_email_address = get_user_email_address(event_type, data)
+
+        # If the event is a group event, get the user email address by
+        # getting all members of the group, and then checking the
+        # verification status of the user with the highest access
+        # level in the group.
+        elif event_type in group_events:
+            logging.info(
+                f"{event_type} event type received, getting user email from GitLab API"
+            )
+            max_access_level = 0
+            user_id_with_max_access = None
+            group_id = data.get("group_id")
+
+            # Get all members of the group
+            response = requests.get(
+                f"{self.gitlab_url}/api/v4/groups/{group_id}/members/all",
+                headers={"PRIVATE-TOKEN": f"{self.gitlab_access_token}"},
+            )
+
+            if response.status_code == 200:
+                gitlab_api_calls_total.labels("success").inc()
+            else:
+                gitlab_api_calls_total.labels("failure").inc()
+
+            try:
+                group_members = response.json()
+            except ValueError:
+                logging.debug(
+                    "Failed to decode JSON from response"
+                )
+                return
+
+            if not isinstance(group_members, list):
+                logging.debug("Unexpected response from server")
+                return
+
+            # Get the user with the highest access level in the group
+            for member in group_members:
+                access_level = member.get("access_level")
+                if access_level > max_access_level:
+                    max_access_level = access_level
+                    user_id_with_max_access = member.get("id")
+                    user_email_address = member.get("email")
+
+            # If no group member with the highest access level had no
+            # public email address present in the list of group members,
+            # get it from their GitLab user attributes.
+            if not user_email_address and user_id_with_max_access is not None:
+                response = requests.get(
+                    f"{self.gitlab_url}/api/v4/users/{user_id_with_max_access}",
+                    headers={"PRIVATE-TOKEN": f"{self.gitlab_access_token}"},
+                )
+                user = response.json()
+                user_email_address = user.get("email")
+
+        # If an email address is still not located and the event type
+        # is not snippet_check, log the situation and return.
+        if user_email_address is None and event_type != "snippet_check":
```

Comment:
**suggestion (non-blocking):** I'd reorder this if
```
if event_type == "snippet_check":
	[...]
elif user_email_address is None:
	[...]
else:
	[...]
```
that removes some of the checks